### ISSUE TITLE
about adagrad whether there exists a error

### ISSUE THREAD

**Issue by chengxiaohua1105** [PaddlePaddle/Paddle#4184]
The adagrad algorithm is implemented as follows:
```
for (size_t i = 0; i < param.size(); ++i) {                        
    accum_g[i] += grad[i] * grad[i];                                 
    param[i] += learning_rate * grad[i] / std::sqrt(accum_g[i] + epsilon_)  + learning_rate * decay_ * param[i];                  
}
```
I think it should be implemented as follows 
```
for (size_t i = 0; i < param.size(); ++i) {                        
    accum_g[i] += grad[i] * grad[i];                                 
    param[i] -= learning_rate * grad[i] / std::sqrt(accum_g[i] + epsilon_) + learning_rate * decay_ * param[i];                  
}
```
"param[i]"  shouls follows "-=" not "+="?

**Comment by Xreki** [comment]
Would you please give me some references?

**Comment by pengli09** [comment]
By default, PaddlePaddle minimizes a loss. So it should be "-=" instead of "+=".

**Comment by chengxiaohua1105** [comment]
tks @pengli09 . I will write pr to fix the bug.
